_posts/2018-11-16-grpc-dotnet-build.md

```diff
@@ -0,0 +1,203 @@
+we won't need, so remove it. Also, we need some `.proto` files to compile. For
+this exercise, we'll copy an example file [`examples/protos/helloworld.proto`
+](https://github.com/grpc/grpc/blob/master/examples/protos/helloworld.proto)
+from the gRPC distribution.
+
+```
+~/work/MyGreeter$ rm Class1.cs
+~/work/MyGreeter$ wget -q https://raw.githubusercontent.com/grpc/grpc/master/examples/protos/helloworld.proto
+```
+
+(on Windows, use `del Class1.cs`, and, if you do not have the wget command,
+just [open the above URL
+](https://raw.githubusercontent.com/grpc/grpc/master/examples/protos/helloworld.proto)
+and use a *Save As...* command from your Web browser).
+
+Next, add required NuGet packages to the project:
+
+```
+~/work/MyGreeter$ dotnet add package Grpc
+info : PackageReference for package 'Grpc' version '1.17.0' added to file '/home/kkm/work/MyGreeter/MyGreeter.csproj'.
+~/work/MyGreeter$ dotnet add package Grpc.Tools
+info : PackageReference for package 'Grpc.Tools' version '1.17.0' added to file '/home/kkm/work/MyGreeter/MyGreeter.csproj'.
+~/work/MyGreeter$ dotnet add package Google.Protobuf
+info : PackageReference for package 'Google.Protobuf' version '3.6.1' added to file '/home/kkm/work/MyGreeter/MyGreeter.csproj'.
+```
+
+### Add `.proto` files to the project
+
+**Next comes an important part.** First of all, by default, a `.csproj` project
+file automatically finds all `.cs` files in its directory, although
+[Microsoft now recommends suppressing this globbing
+behavior](https://docs.microsoft.com/dotnet/core/tools/csproj#recommendation),
+so we too decided against globbing `.proto` files. Thus the `.proto`
+files must be added to the project explicitly.
+
+Second of all, it is important to add a property `PrivateAssets="All"` to the
+Grpc.Tools package reference, so that it will not be needlessly fetched by the
+consumers of your new library. This makes sense, as the package only contains
+compilers, code generators and import files, which are not needed outside of
+the project where the `.proto` files have been compiled. While not strictly
+required in this simple walkthrough, it must be your standard practice to do
+that always.
+
+So edit the file `MyGreeter.csproj` to add the `helloworld.proto` so that it
+will be compiled, and the `PrivateAssets` property to the Grpc.Tools package
+reference. Your resulting project file should now look like this:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Google.Protobuf" Version="3.6.1" />
+    <PackageReference Include="Grpc" Version="1.17.0" />
+                                                     <!--  Add attribute as shown below. -->
+    <PackageReference Include="Grpc.Tools" Version="1.17.0" PrivateAssets="All" />
+    <Protobuf Include="helloworld.proto" />  <!-- Add this entire line. -->
```

Comment:
nit: I'd still rephrase the comment to carry more meaning.
```
<!-- use the Grpc.Tools package to autogenerate C# sources from .proto file -->
<Protobuf Include="helloworld.proto" />
```

Reply:
Done